casper/src/main/scala/coop/rchain/casper/Estimator.scala

```diff
@@ -190,17 +180,16 @@ object Estimator {
       blockDag: BlockDagRepresentation[F],
       scores: Map[BlockHash, Long]
   ): F[List[BlockHash]] =
-    for {
-      children <- blockDag
-                   .children(b)
-                   .map(maybeChildren => maybeChildren.getOrElse(Set.empty[BlockHash]))
-      scoredChildren = children.filter(scores.contains)
-      result = if (scoredChildren.nonEmpty) {
-        scoredChildren.toList
-      } else {
-        List(b)
-      }
-    } yield result
+    blockDag.children(b).map {
+      case Some(children) =>
+        val scoredChildren = children.filter(scores.contains)
+        if (scoredChildren.nonEmpty) {
+          scoredChildren.toList
+        } else {
+          List(b)
+        }
+      case None => List(b)
+    }
```

Comment:
Trying to understand what happens here I got to the following refactoring which I find more readable (but I don't have too strong opinions):

```scala
 private  def toNonEmptyList[T](elements  :Set[T]) : Option[List[T]] =
    if (elements.isEmpty) None
    else Some(elements.toList)

  private def replaceBlockHashWithChildren[F[_]: Monad](
      b: BlockHash,
      blockDag: BlockDagRepresentation[F],
      scores: Map[BlockHash, Long]
  ): F[List[BlockHash]] =
    blockDag.children(b)
      .map (
       _.flatMap { children => toNonEmptyList(children.filter(scores.contains))}
        .getOrElse(List(b))
      )
```

Reply:
What about

```
  private def toNonEmptyList[T](elements: Set[T]) : Option[List[T]] =
    if (elements.isEmpty) None
    else Some(elements.toList)

  private def replaceBlockHashWithChildren[F[_]: Monad](
                                                         b: BlockHash,
                                                         blockDag: BlockDagRepresentation[F],
                                                         scores: Map[BlockHash, Long]
                                                       ): F[List[BlockHash]] =
    blockDag.children(b).map {
      case Some(children) =>
        val scoredChildren = children.filter(scores.contains)
        toNonEmptyList(scoredChildren).getOrElse(List(b))
      case None =>
        List(b)
    }
```